File: Keepr.Server/Services/VaultsService.cs
using System;
using System.Collections.Generic;
using Keepr.Server.Models;
using Keepr.Server.Repositories;

namespace Keepr.Server.Services
{
    public class VaultsService
    {
        private readonly VaultsRepository _repo;
    private readonly VaultKeepsRepository _repoVK;

    public VaultsService(VaultsRepository vs, VaultKeepsRepository vkr)
        {
            _repo = vs;
            _repoVK = vkr;
        }

    // internal List<Vault> GetAll()
    // {
    //   return _repo.GetAll();
    // }

    internal Vault GetById(int id)
    {
      Vault vault = _repo.GetById(id);
      if(vault == null)
      {
        throw new Exception("Invalid ID");
      }
      return vault;
    }

    internal List<VaultKeepViewModel> GetVaultKeeps(int id)
    {
      return _repoVK.GetVaultKeeps(id);
    }

    internal List<Vault> GetVaultsByAccountId(string id)
    {
      return _repo.GetVaultsByAccountId(id);
    }

    internal List<Vault> GetVaultsByProfileId(string id)
    {
      return _repo.GetVaultsByProfileId(id);
    }

    internal Vault Create(Vault v)
    {
      return _repo.Create(v);
    }

    internal Vault Edit(Vault v, string id)
    {
      Vault original = _repo.GetById(v.Id);
      if(original == null)
      {
        throw new Exception("Invalid Id");
      }
      if(original.CreatorId != id)
      {
        throw new Exception("You cannot edit this");
      }
      v.CreatorId = id;
      return _repo.Edit(v);
    }
    

    internal void Delete(int id, string creatorId)
    {
      Vault vault = GetById(id);
      if(vault.CreatorId != creatorId)
      {
        throw new Exception("You cannot delete another user's blog");
      }
      _repo.Delete(id);
    }


  }
}

File: Keepr.Client/src/services/VaultsService.js
import { AppState } from '../AppState'
import { api } from './AxiosService'

class VaultsService {
  async getById(id) {
    const res = await api.get(`api/vaults/${id}`)
    AppState.activeVault = res.data
  }

  async getVaultsByProfileId(id) {
    const res = await api.get(`api/profiles/${id}/vaults`)
    AppState.vaults = res.data.reverse()
  }

  async getVaultsByAccountId(id) {
    const res = await api.get(`api/profiles/${id}/vaults`)
    AppState.accountVaults = res.data.reverse()
  }

  async create(newVault) {
    const res = await api.post('api/vaults', newVault)
    this.getVaultsByAccountId(res.data.creatorId)
  }

  async edit(vault) {
    await api.put(`api/vaults/${vault.id}`, vault)
    this.getVaultsByProfileId(vault.creatorId)
  }

  async delete(id) {
    await api.delete(`api/vaults/${id}`)
    this.getVaultsByAccountId(AppState.account.id)
  }
}
export const vaultsService = new VaultsService()
